File: plugins/streamlit-to-sigma/skills/streamlit-to-sigma/scripts/cleanup-orphan-workbooks.rb
#!/usr/bin/env ruby
# Review and, only with per-workbook user confirmation, delete orphan Sigma
# workbooks left by spec-iteration retries.
#
# SAFETY CONTRACT:
#   * --keep is mandatory when the ledger contains more than one workbook.
#     The last ledger line is NOT assumed to be live: workdirs and their
#     append-only ledgers can be reused across migration runs.
#   * Every candidate is read back from BOTH the workbook and file APIs before
#     deletion. It must resolve to the same workbook inode, folder, creator, and
#     owner context as the explicitly kept workbook.
#   * Deletion is interactive only. The operator must type each candidate's full
#     ID after reviewing its live name/path/timestamps. There is deliberately no
#     --yes/non-interactive bypass.
#   * --dry-run performs all read-only validation and writes the proposed IDs,
#     but never prompts or deletes.
#
# Usage:
#   ruby scripts/cleanup-orphan-workbooks.rb --workdir /tmp/<name> --keep <live-id>
#   ruby scripts/cleanup-orphan-workbooks.rb --workdir /tmp/<name> --keep <live-id> --dry-run
#
# This script is NOT a way to delete the workbook just built. It always preserves
# --keep. Use Sigma's UI when the current workbook itself should be removed.
#
# Writes <workdir>/cleanup-marker.json for assert-phase6-ran.rb.
#
# Exit codes:
#   0  no cleanup needed, dry-run completed, or all candidates were confirmed/deleted
#   1  a candidate was refused, skipped, or failed
#   2  unsafe/invalid invocation (including a non-interactive destructive run)

require 'net/http'
require 'uri'
require 'json'
require 'time'
require 'optparse'

module OrphanWorkbookCleanup
  ID_PATTERN = /\A[A-Za-z0-9_-]+\z/.freeze

  module_function

  def parse_options(argv)
    opts = { dry_run: false }
    OptionParser.new do |parser|
      parser.on('--workdir DIR') { |value| opts[:workdir] = value }
      parser.on('--dry-run') { opts[:dry_run] = true }
      parser.on('--keep ID') { |value| opts[:keep] = value }
      # Retained for CLI compatibility. A missing ledger is always a safe no-op.
      parser.on('--allow-empty-log') { opts[:allow_empty] = true }
    end.parse!(argv)
    opts
  end

  def load_ledger(path)
    records = []
    File.readlines(path).each_with_index do |line, index|
      next if line.strip.empty?
      begin
        record = JSON.parse(line)
      rescue JSON::ParserError => e
        raise ArgumentError, "invalid JSON at #{File.basename(path)} line #{index + 1}: #{e.message}"
      end
      unless record.is_a?(Hash) && record['id'].is_a?(String) &&
             !record['id'].empty? && ID_PATTERN.match?(record['id'])
        raise ArgumentError, "#{File.basename(path)} line #{index + 1} must contain a safe string id"
      end
      records << record
    end
    records
  end

  def parse_json_response(response, label)
    code = response.code.to_i
    raise "#{label} returned HTTP #{code}" unless code.between?(200, 299)
    parsed = JSON.parse(response.body.to_s)
    raise "#{label} returned a non-object body" unless parsed.is_a?(Hash)
    parsed
  rescue JSON::ParserError => e
    raise "#{label} returned invalid JSON: #{e.message}"
  end

  def inspect_workbook(id, requester)
    workbook = parse_json_response(requester.call(:get, "/v2/workbooks/#{id}"),
                                   "GET /v2/workbooks/#{id}")
    file = parse_json_response(requester.call(:get, "/v2/files/#{id}"),
                               "GET /v2/files/#{id}")

    raise "workbook API returned #{workbook['workbookId'].inspect}, expected #{id}" unless workbook['workbookId'] == id
    raise "file API returned #{file['id'].inspect}, expected #{id}" unless file['id'] == id
    raise "file #{id} is type #{file['type'].inspect}, not workbook" unless file['type'] == 'workbook'
    if workbook['workbookUrlId'] && file['urlId'] && workbook['workbookUrlId'] != file['urlId']
      raise "workbook/file URL identifiers disagree for #{id}"
    end

    {
      'id' => id,
      'name' => file['name'] || workbook['name'],
      'path' => file['path'] || workbook['path'],
      'parentId' => file['parentId'],
      'ownerId' => file['ownerId'] || workbook['ownerId'],
      'createdBy' => file['createdBy'] || workbook['createdBy'],
      'createdAt' => file['createdAt'] || workbook['createdAt'],
      'updatedAt' => file['updatedAt'] || workbook['updatedAt']
    }
  end

  def same_cleanup_context?(candidate, kept)
    %w[parentId ownerId createdBy].all? do |key|
      candidate[key] && kept[key] && candidate[key] == kept[key]
    end
  end

  def marker_path(workdir)
    File.join(workdir, 'cleanup-marker.json')
  end

  def write_marker(workdir, marker)
    File.write(marker_path(workdir), JSON.pretty_generate(marker))
  end

  def tty?(input)
    input.respond_to?(:tty?) && input.tty?
  end

  def run(argv, env: ENV, input: $stdin, output: $stdout, error: $stderr, requester: nil)
    opts = parse_options(argv.dup)
    unless opts[:workdir]
      error.puts '--workdir required'
      return 2
    end

    log = File.join(opts[:workdir], 'posted-workbooks.jsonl')
    unless File.exist?(log)
      output.puts "[OK] no posted-workbooks.jsonl at #{log} — nothing to clean up"
      return 0
    end

    begin
      records = load_ledger(log)
    rescue ArgumentError => e
      error.puts "REFUSED: #{e.message}"
      return 2
    end
    if records.empty?
      output.puts '[OK] posted-workbooks.jsonl is empty — nothing to clean up'
      return 0
    end

    unique_ids = records.map { |record| record['id'] }.uniq
    if unique_ids.length == 1
      if opts[:keep] && opts[:keep] != unique_ids.first
        error.puts "REFUSED: --keep #{opts[:keep]} is not the ledger workbook #{unique_ids.first}"
        return 2
      end
      keep_id = opts[:keep] || unique_ids.first
      output.puts "[OK] only one POSTed workbook (#{keep_id}) — no orphans to clean up"
      write_marker(opts[:workdir], {
                     'ran_at' => Time.now.utc.iso8601, 'kept' => keep_id,
                     'deleted' => [], 'failed' => [], 'skipped' => [],
                     'dry_run' => opts[:dry_run]
                   })
      return 0
    end

    keep_id = opts[:keep].to_s
    if keep_id.empty?
      error.puts 'REFUSED: --keep <live-workbook-id> is required; ledger order is not proof of which workbook is live.'
      error.puts "Ledger IDs: #{unique_ids.join(', ')}"
      return 2
    end
    unless ID_PATTERN.match?(keep_id) && unique_ids.include?(keep_id)
      error.puts "REFUSED: --keep #{keep_id.inspect} is not a safe ID present in posted-workbooks.jsonl"
      return 2
    end
    unless opts[:dry_run] || tty?(input)
      error.puts 'REFUSED: workbook deletion requires an interactive terminal.'
      error.puts "Review first: ruby #{__FILE__} --workdir #{opts[:workdir]} --keep #{keep_id} --dry-run"
      error.puts 'Then run the same command without --dry-run and type each workbook ID when prompted.'
      return 2
    end

    unless requester
      base = env['SIGMA_BASE_URL']
      unless base && !base.empty?
        error.puts 'SIGMA_BASE_URL not set'
        return 2
      end
      requester = real_requester(base, env)
    end

    begin
      kept = inspect_workbook(keep_id, requester)
    rescue StandardError => e
      error.puts "REFUSED: could not validate the workbook to keep: #{e.message}"
      return 2
    end

    candidates = unique_ids.reject { |id| id == keep_id }
    output.puts "KEEPING (explicit --keep): #{kept['name'].inspect}"
    output.puts "  id:      #{keep_id}"
    output.puts "  path:    #{kept['path']}"
    output.puts "  created: #{kept['createdAt']}"
    output.puts
    output.puts "Reviewing #{candidates.length} ledger candidate(s). Ledgers can span multiple runs."
    output.puts 'Confirm only workbooks you recognize as retries from this migration.'

    deleted = []
    failed = []
    skipped = []
    would_delete = []

    candidates.each do |id|
      begin
        candidate = inspect_workbook(id, requester)
        unless same_cleanup_context?(candidate, kept)
          raise 'folder, owner, or creator differs from the explicitly kept workbook'
        end
      rescue StandardError => e
        error.puts "  [REFUSED] #{id}: #{e.message}"
        failed << { 'id' => id, 'reason' => e.message }
        next
      end

      output.puts
      output.puts 'CANDIDATE:'
      output.puts "  name:    #{candidate['name'].inspect}"
      output.puts "  id:      #{id}"
      output.puts "  path:    #{candidate['path']}"
      output.puts "  created: #{candidate['createdAt']}"
      output.puts "  updated: #{candidate['updatedAt']}"

      if opts[:dry_run]
        output.puts '  [DRY-RUN] validated; no deletion performed'
        would_delete << candidate
        next
      end

      output.print "Type the full workbook ID to delete this candidate, or press Enter to keep it:\n> "
      output.flush
      confirmation = input.gets.to_s.strip
      unless confirmation == id
        output.puts "  [KEPT] #{id} — confirmation did not match"
        skipped << { 'id' => id, 'reason' => 'user did not type the full workbook ID' }
        next
      end

      response = requester.call(:delete, "/v2/files/#{id}")
      code = response.code.to_i
      if code.between?(200, 299) || code == 404
        output.puts "  [deleted] #{id} (HTTP #{code})"
        deleted << { 'id' => id, 'status' => code }
      else
        body = response.body.to_s[0..200]
        error.puts "  [FAIL] #{id} (HTTP #{code}) — #{body}"
        failed << { 'id' => id, 'status' => code, 'body' => body }
      end
    end

    marker = {
      'ran_at' => Time.now.utc.iso8601,
      'kept' => keep_id,
      'deleted' => deleted,
      'failed' => failed,
      'skipped' => skipped,
      'dry_run' => opts[:dry_run]
    }
    marker['would_delete'] = would_delete if opts[:dry_run]
    write_marker(opts[:workdir], marker)

    if opts[:dry_run]
      output.puts "\n[DRY-RUN] #{would_delete.length} candidate(s) validated; no DELETE calls made."
      return failed.empty? ? 0 : 1
    end
    if failed.any? || skipped.any?
      error.puts "\nCleanup incomplete: deleted #{deleted.length}, refused/failed #{failed.length}, kept by user #{skipped.length}."
      return 1
    end

    output.puts "\n[OK] user confirmed and deleted #{deleted.length} orphan workbook(s); kept #{keep_id}"
    0
  rescue OptionParser::ParseError => e
    error.puts e.message
    2
  end

  def real_requester(base, env)
    $LOAD_PATH.unshift File.expand_path('lib', __dir__)
    require 'sigma_rest'
    lambda do |method, path|
      attempts = 0
      loop do
        attempts += 1
        uri = URI("#{base}#{path}")
        request = method == :delete ? Net::HTTP::Delete.new(uri) : Net::HTTP::Get.new(uri)
        request['Authorization'] = "Bearer #{Sigma.auth_token}"
        request['Accept'] = 'application/json'
        response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https',
                                  read_timeout: 30) { |http| http.request(request) }
        if response.code.to_i == 401 && attempts == 1 && env['SIGMA_CLIENT_ID']
          Sigma.refresh_token!
          next
        end
        break response
      end
    end
  end
end

exit OrphanWorkbookCleanup.run(ARGV) if $PROGRAM_NAME == __FILE__

File: plugins/qlik-to-sigma/skills/qlik-to-sigma/scripts/cleanup_orphan_workbooks.py
#!/usr/bin/env python3
"""Safely review and delete retry workbooks recorded in a migration ledger."""

from __future__ import annotations

import argparse
import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable

from lib import sigma_rest

ID_PATTERN = re.compile(r"^[A-Za-z0-9_-]+$")


def now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace(
        "+00:00", "Z"
    )


def load_ledger(path: Path) -> list[dict[str, Any]]:
    records = []
    for index, line in enumerate(path.read_text(encoding="utf-8-sig").splitlines(), 1):
        if not line.strip():
            continue
        try:
            record = json.loads(line)
        except json.JSONDecodeError as exc:
            raise ValueError(f"{path.name} line {index}: {exc}") from exc
        if (
            not isinstance(record, dict)
            or not isinstance(record.get("id"), str)
            or not ID_PATTERN.fullmatch(record["id"])
        ):
            raise ValueError(
                f"{path.name} line {index} must contain a safe string id"
            )
        records.append(record)
    return records


def inspect_workbook(
    workbook_id: str, requester: Callable[[str, str], Any]
) -> dict[str, Any]:
    workbook = requester("get", f"/v2/workbooks/{workbook_id}") or {}
    file_row = requester("get", f"/v2/files/{workbook_id}") or {}
    if workbook.get("workbookId") != workbook_id:
        raise ValueError(
            f"workbook API returned {workbook.get('workbookId')!r}, "
            f"expected {workbook_id}"
        )
    if file_row.get("id") != workbook_id:
        raise ValueError(
            f"file API returned {file_row.get('id')!r}, expected {workbook_id}"
        )
    if file_row.get("type") != "workbook":
        raise ValueError(
            f"file {workbook_id} is type {file_row.get('type')!r}, not workbook"
        )
    if (
        workbook.get("workbookUrlId")
        and file_row.get("urlId")
        and workbook["workbookUrlId"] != file_row["urlId"]
    ):
        raise ValueError(f"workbook/file URL identifiers disagree for {workbook_id}")
    return {
        "id": workbook_id,
        "name": file_row.get("name") or workbook.get("name"),
        "path": file_row.get("path") or workbook.get("path"),
        "parentId": file_row.get("parentId"),
        "ownerId": file_row.get("ownerId") or workbook.get("ownerId"),
        "createdBy": file_row.get("createdBy") or workbook.get("createdBy"),
        "createdAt": file_row.get("createdAt") or workbook.get("createdAt"),
        "updatedAt": file_row.get("updatedAt") or workbook.get("updatedAt"),
    }


def same_context(candidate: dict[str, Any], kept: dict[str, Any]) -> bool:
    return all(
        candidate.get(key)
        and kept.get(key)
        and candidate[key] == kept[key]
        for key in ("parentId", "ownerId", "createdBy")
    )


def write_marker(workdir: Path, value: dict[str, Any]) -> None:
    (workdir / "cleanup-marker.json").write_text(
        json.dumps(value, indent=2) + "\n", encoding="utf-8"
    )


def cleanup(
    workdir: Path,
    keep_id: str | None,
    *,
    dry_run: bool,
    requester: Callable[[str, str], Any],
    input_stream: Any = sys.stdin,
    output_stream: Any = sys.stdout,
    error_stream: Any = sys.stderr,
) -> int:
    ledger_path = workdir / "posted-workbooks.jsonl"
    if not ledger_path.is_file():
        print(
            f"[OK] no posted-workbooks.jsonl at {ledger_path} — nothing to clean up",
            file=output_stream,
        )
        return 0
    try:
        records = load_ledger(ledger_path)
    except (OSError, ValueError) as exc:
        print(f"REFUSED: {exc}", file=error_stream)
        return 2
    unique_ids = list(dict.fromkeys(row["id"] for row in records))
    if not unique_ids:
        print("[OK] posted-workbooks.jsonl is empty — nothing to clean up", file=output_stream)
        return 0
    if len(unique_ids) == 1:
        if keep_id and keep_id != unique_ids[0]:
            print(
                f"REFUSED: --keep {keep_id} is not the ledger workbook {unique_ids[0]}",
                file=error_stream,
            )
            return 2
        kept = keep_id or unique_ids[0]
        write_marker(
            workdir,
            {
                "ran_at": now(),
                "kept": kept,
                "deleted": [],
                "failed": [],
                "skipped": [],
                "dry_run": dry_run,
            },
        )
        print(
            f"[OK] only one POSTed workbook ({kept}) — no orphans to clean up",
            file=output_stream,
        )
        return 0
    if (
        not keep_id
        or not ID_PATTERN.fullmatch(keep_id)
        or keep_id not in unique_ids
    ):
        print(
            "REFUSED: --keep must be a safe ID present in posted-workbooks.jsonl",
            file=error_stream,
        )
        return 2
    if not dry_run and not getattr(input_stream, "isatty", lambda: False)():
        print(
            "REFUSED: workbook deletion requires an interactive terminal.",
            file=error_stream,
        )
        return 2
    try:
        kept = inspect_workbook(keep_id, requester)
    except (ValueError, sigma_rest.SigmaError) as exc:
        print(
            f"REFUSED: could not validate the workbook to keep: {exc}",
            file=error_stream,
        )
        return 2
    print(f"KEEPING (explicit --keep): {kept.get('name')!r}", file=output_stream)
    deleted, failed, skipped, would_delete = [], [], [], []
    for workbook_id in (value for value in unique_ids if value != keep_id):
        try:
            candidate = inspect_workbook(workbook_id, requester)
            if not same_context(candidate, kept):
                raise ValueError(
                    "folder, owner, or creator differs from the kept workbook"
                )
        except (ValueError, sigma_rest.SigmaError) as exc:
            print(f"  [REFUSED] {workbook_id}: {exc}", file=error_stream)
            failed.append({"id": workbook_id, "reason": str(exc)})
            continue
        print(
            f"\nCANDIDATE:\n  name: {candidate.get('name')!r}\n"
            f"  id: {workbook_id}\n  path: {candidate.get('path')}\n"
            f"  created: {candidate.get('createdAt')}\n"
            f"  updated: {candidate.get('updatedAt')}",
            file=output_stream,
        )
        if dry_run:
            print("  [DRY-RUN] validated; no deletion performed", file=output_stream)
            would_delete.append(candidate)
            continue
        print(
            "Type the full workbook ID to delete this candidate, "
            "or press Enter to keep it:\n> ",
            end="",
            file=output_stream,
            flush=True,
        )
        if input_stream.readline().strip() != workbook_id:
            skipped.append(
                {"id": workbook_id, "reason": "user did not type the full workbook ID"}
            )
            print(f"  [KEPT] {workbook_id}", file=output_stream)
            continue
        try:
            requester("delete", f"/v2/files/{workbook_id}")
            deleted.append({"id": workbook_id, "status": 200})
            print(f"  [deleted] {workbook_id}", file=output_stream)
        except sigma_rest.SigmaError as exc:
            if "-> 404" in str(exc).splitlines()[0]:
                deleted.append({"id": workbook_id, "status": 404})
                print(
                    f"  [deleted] {workbook_id} (already absent)",
                    file=output_stream,
                )
            else:
                failed.append({"id": workbook_id, "reason": str(exc)})
                print(f"  [FAIL] {workbook_id}: {exc}", file=error_stream)
    marker = {
        "ran_at": now(),
        "kept": keep_id,
        "deleted": deleted,
        "failed": failed,
        "skipped": skipped,
        "dry_run": dry_run,
    }
    if dry_run:
        marker["would_delete"] = would_delete
    write_marker(workdir, marker)
    return 0 if not failed and (dry_run or not skipped) else 1


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--workdir", required=True)
    parser.add_argument("--keep")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--allow-empty-log", action="store_true")
    args = parser.parse_args(argv)
    return cleanup(
        Path(args.workdir).expanduser().resolve(),
        args.keep,
        dry_run=args.dry_run,
        requester=lambda method, path: sigma_rest.request(method, path),
    )


if __name__ == "__main__":
    raise SystemExit(main())
